### PR TITLE
[jit] speed up alias analysis

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37106 [wip] update graph fuser aliasdb in-place
* **#36345 [jit] speed up alias analysis**

During compilation, we spend a huge amount of time in alias analyis.
This PR does a few things to speed it up.

1. Separate the analysis into two phases: one where we build up the
necessary data structures, and the other where we service aliasing
queries. This allows us to defer building indices/maintaining index
consistency until after the "buildup" phase is done.

2. Properly memoize/dynamic program the memory locations lookups.

3. Done naively, setting wildcards invalidates the above memoization,
trigger costly recomputation. So I added a cache-aware `setWildcards`.
Sadly that means you need alias analysis to reach into the guts of
memorydag, but the speedup is worth it.

Sadly, these changes are kind of coupled for correctness reasons, so
they're all here at once.

I used this model (thanks @IlyaOvodov) as a provisional benchmark. You
can get it here:
https://www.dropbox.com/s/jlyygn6yygj1jkx/yolov3.zip. Unzip at run
`python test_timing.py`.

Baseline: (752.076s) right before 6bc8ffe82462c77ac4f9b27452046cb1f8f07d92
After optimizing before inlining: (699.593s)
After deferring cache construction: (426.180s)
After cache-aware `setWildcards`: (193.678s)

So a nice 75% speedup to overall compilation. There's a lot more to do
in other places of the compilation pipeline though.

Followup to this PR specifically:  Everything that fans out from the
`analyze` call is the "buildup" phase of AliasDB construction. This
should be factored into a separate analysis pass to statically
distinguish the two phases (right now we just null out stuff to
accomplish the same thing dynamically).

Differential Revision: [D20952727](https://our.internmc.facebook.com/intern/diff/D20952727)

### CODE REVIEW

**Comment by eellison** [comment]
@suo could you look into the failures before I review? i would expect this shouldn't have any behavioral changes bc it's not changing the analysis just speeding it up (it shouldn't break tests).

**Comment by IlyaOvodov** [comment]
@suo, 75% performance boosting is a great success, but still an application that starts for 3 minutes is almost as useless as one that starts for 10 min.
As far as I understood from reading issues and forums, this huge time is spent on optimization (that gives about several % of performance boost and is repeated every time when input shape is changed), compared with pure python PyTorch which starts the same net in less than 1 sec.
If I'm right, possibly it's a solution to give a way to switch this optimisation off when time it takes become unacceptable? Sorry if the reality is more complex than I expect :)

**Comment by suo** [comment]
@IlyaOvodov: correct, the optimization time on large networks is still way too high. In production environments, we typically expect this time to be amortized over the lifetime of model serving, but even then the current state is bad.

There are a number of ways to tune the execution strategy. For example, if you don't want to optimization and basically want the code to run in interpreter-mode, you can set the following:
```
with torch.jit.optimized_execution(false):
    my_model(inputs)
```

**Comment by IlyaOvodov** [comment]
@suo Wow! What a great undocumented cheat code! 
But is there a way to do the same (turn it off) in C++ inference?

**Comment by suo** [comment]
> But is there a way to do the same (turn it off) in C++ inference?
answered on https://github.com/pytorch/pytorch/issues/36040 as it's more appropriate for discussion.

**Comment by eellison** [comment]
@suo i'm reviewing now but could you expand parts 2 and 3 in your summary with more information?
> Properly memoize/dynamic program the memory locations lookups.
how ?

> Done naively, setting wildcards invalidates the above memoization,
trigger costly recomputation. So I added a cache-aware setWildcards.
Sadly that means you need alias analysis to reach into the guts of
memorydag, but the speedup is worth it.
what is the naive solution and why doesn't it work, what is the new approach why does that solve it

**Comment by suo** [comment]
> @suo i'm reviewing now but could you expand parts 2 and 3 in your summary with more information?
> 
> > Properly memoize/dynamic program the memory locations lookups.

The previous bfs implementation wasn't re-using the cached memorylocation lookup result, so we would re-do the whole search instead of memoizing.


> > Done naively, setting wildcards invalidates the above memoization,
> > trigger costly recomputation. So I added a cache-aware setWildcards.
> > Sadly that means you need alias analysis to reach into the guts of
> > memorydag, but the speedup is worth it.
> 
> what is the naive solution and why doesn't it work, what is the new approach why does that solve it

Naive solution is to invalidate everything in the transitive "points-from" closure to the wildcard. Since `setWildcard` only affects memorylocations (which are at the "source" of the points-from graph), it has the effect of invalidating basically the entire cache.

The new approach is commented in the code, but basically involves updating the cache in place in linear time.

File: torch/csrc/jit/ir/alias_analysis.cpp [hunk 1]
```diff
@@ -126,10 +126,78 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
...
+
+  for (const auto& write : writeRegistry_->containedWrites_) {
+    auto elem = elementMap_.at(write.second);
+    MemoryLocations mem_locations;
+    memoryDAG_->collectAllContainedMemoryLocations(elem, mem_locations);
+    for (unsigned loc : mem_locations) {
+      auto contained_elem = memoryDAG_->fromIndex(loc);
+      // we only register writes on memory locations
+      if (contained_elem->pointsTo.empty()) {
+        writeIndex[write.first].set(contained_elem->index);
```
Comment (eellison) [inline torch/csrc/jit/ir/alias_analysis.cpp:185]: Why are we calling `.set()` here and `|=` above ? Could we be just be using `=` instead of `|=` above

File: torch/csrc/jit/ir/alias_analysis.cpp [hunk 2]
```diff
@@ -126,10 +126,78 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
...
+      }
+    }
+  }
+
+  for (const auto& write : writeRegistry_->writesToAllWildcards_) {
```
Comment (eellison) [inline torch/csrc/jit/ir/alias_analysis.cpp:190]: I'm not sure why we have `writesToAllWildcards`. It's only called on `aten::wait`, where we could instead just `analyzeConservative`.

File: torch/csrc/jit/ir/alias_analysis.cpp [hunk 3]
```diff
@@ -173,16 +241,13 @@ bool AliasDb::hasWriters(const Value* v) const {
     return false;
   }
```
Comment (eellison) [inline torch/csrc/jit/ir/alias_analysis.cpp:242]: Unrelated to this PR but: in `hasWriters` we're not testing the invariant that mutable values have an `Element *`

File: torch/csrc/jit/ir/alias_analysis.cpp [hunk 4]
```diff
@@ -1409,38 +1467,21 @@ c10::optional<Element*> AliasDb::setWildcard(const Value* v) {
...
-  isWriteCacheStale_ = true;
-  return wildcardElement;
+  return *maybe_wildcardElement;
 }
 
-void AliasDb::rebuildWriteCache() const {
-  writeCache_ = {};
-  for (const auto& pr : writeIndex_) {
+MemoryLocations AliasDb::buildWriteCache() const {
+  MemoryLocations ret;
```
Comment (eellison) [inline torch/csrc/jit/ir/alias_analysis.cpp:1478]: Why was this changed from mutating & void to functional and returning a new value?

File: torch/csrc/jit/passes/utils/memory_dag.cpp [hunk 5]
```diff
@@ -93,71 +107,103 @@ bool MemoryDAG::mayContainAlias(
...
+    // Base case: if we don't point to anything, this element is a memory location.
+    // Return itself.
+    ret.set(e->index);
+  } else {
+    for (auto el : e->pointsTo) {
+      ret |= getMemoryLocations(fromIndex(el));
+    }
+  }
+
+  e->cachedMemoryLocations_ = std::move(ret);
```
Comment (eellison) [inline torch/csrc/jit/passes/utils/memory_dag.cpp:145]: What prevents this cache from getting invalidated ?

**Review by eellison — COMMENTED** [review]
I had a hard time reviewing this code for correctness. There are a lot of new caches, and it's hard to parse out what their lifetimes are and when they are invalidated etc. 

Could you please write up a description of how everything works together?

File: torch/csrc/jit/passes/utils/memory_dag.cpp (hunk 5 above)
Comment (suo) [reply torch/csrc/jit/passes/utils/memory_dag.cpp:145]: MemoryDAG is immutable (except through the specially marked `setWildcard` which is responsible for keeping the caches consistent).

File: torch/csrc/jit/ir/alias_analysis.cpp (hunk 4 above)
Comment (suo) [reply torch/csrc/jit/ir/alias_analysis.cpp:1478]: Just because it is easier for me to reason about functional things

File: torch/csrc/jit/ir/alias_analysis.cpp (hunk 2 above)
Comment (suo) [reply torch/csrc/jit/ir/alias_analysis.cpp:190]: the comment explains why. We cannot just be conservative and write to all inputs, because we don't know what the forked inputs are.

File: torch/csrc/jit/ir/alias_analysis.h [hunk 6]
```diff
@@ -233,11 +235,18 @@ class AliasDb {
    * State for tracking write info.
    */
   // Map of nodes to the memory locations that they write to
-  ska::flat_hash_map<Node*, MemoryLocations> writeIndex_;
-  // Set of all memory locations that may have been written to.
-  mutable MemoryLocations writeCache_;
-  mutable bool isWriteCacheStale_ = true;
-  void rebuildWriteCache() const;
+  using TWriteIndex = ska::flat_hash_map<Node*, MemoryLocations>;
```
[the remark this answers is not in the document]
Comment (suo) [reply torch/csrc/jit/ir/alias_analysis.h:238]: Type

File: torch/csrc/jit/passes/utils/memory_dag.h [hunk 7]
```diff
@@ -69,22 +68,59 @@ class TORCH_API MemoryDAG {
...
+  void addToContainedElements(Element* contained, Element* container);
+  // Make a fresh element (i.e. an element that doesn't point to anything) and
+  // return it.
+  Element* makeFreshValue(const Value* v);
 
+  // Produce a query-able MemoryDAG from `this`.
+  //
+  // NOTE: This function consumes this builder, and any further mutation/usage
+  // is undefined.
+  std::unique_ptr<MemoryDAG> build();
```
[the remark this answers is not in the document]
Comment (suo) [reply torch/csrc/jit/passes/utils/memory_dag.h:112]: good idea, will change

File: torch/csrc/jit/passes/utils/memory_dag.h [hunk 8]
```diff
@@ -69,22 +68,59 @@ class TORCH_API MemoryDAG {
...
   bool mayContainAliasImpl(const Element* contained, const Element* container)
       const;
+  std::vector<std::unique_ptr<Element>> indexToElementMap_;
+};
+
+/**
+ * Helper to build up the points-to graph.
+ *
+ * We separate the "building" into a different class because it allows us to
+ * cache internally to MemoryDAG without worrying about how the DAG structure
```
[the remark this answers is not in the document]
Comment (suo) [reply torch/csrc/jit/passes/utils/memory_dag.h:95]: It means: during building we do not build caches. Then, when we're done, we can write caches is on an immutable object which is a lot simplier

File: torch/csrc/jit/ir/alias_analysis.cpp [hunk 9]
```diff
@@ -126,10 +126,78 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
...
+    }
+    void registerWriteToAllContained(const Value* v, Node* n) {
+      containedWrites_.emplace(n, v);
+    }
+    void registerWriteToAllWildcards(Node* n) {
+      writesToAllWildcards_.insert(n);
+    }
+    std::unordered_map<Node*, const Value*> writes_;
+    std::unordered_map<Node*, const Value*> containedWrites_;
+    std::unordered_set<Node*> writesToAllWildcards_;
```
[the remark this answers is not in the document]
Comment (suo) [reply torch/csrc/jit/ir/alias_analysis.cpp:143]: explained on aten::wait discussion below

File: torch/csrc/jit/ir/alias_analysis.h [hunk 10]
```diff
@@ -233,11 +235,18 @@ class AliasDb {
...
+  c10::optional<TWriteIndex> writeIndex_;
+
+  struct WriteRegistry;
+  std::unique_ptr<WriteRegistry> writeRegistry_;
+
+  mutable c10::optional<MemoryLocations> writeCache_;
+  MemoryLocations buildWriteCache() const;
+
+  std::unordered_set<const Value*> wildcards_;
+  std::unordered_map<Element*, std::vector<const Value*>> wc_;
```
[the remark this answers is not in the document]
Comment (suo) [reply torch/csrc/jit/ir/alias_analysis.h:248]: yeah actually this is totally redundant and unused but snuck in there :P I'll fix

File: torch/csrc/jit/ir/alias_analysis.h [hunk 11]
```diff
@@ -233,11 +235,18 @@ class AliasDb {
...
-  ska::flat_hash_map<Node*, MemoryLocations> writeIndex_;
-  // Set of all memory locations that may have been written to.
-  mutable MemoryLocations writeCache_;
-  mutable bool isWriteCacheStale_ = true;
-  void rebuildWriteCache() const;
+  using TWriteIndex = ska::flat_hash_map<Node*, MemoryLocations>;
+  c10::optional<TWriteIndex> writeIndex_;
+
+  struct WriteRegistry;
+  std::unique_ptr<WriteRegistry> writeRegistry_;
```
[the remark this answers is not in the document]
Comment (suo) [reply torch/csrc/jit/ir/alias_analysis.h:242]: I will add in a comment:
- writeRegistry_ keeps track of writes as we are building aliasdb
- writeIndex_ is built from the `writeRegistry_` and can answer queries efficiently
- writeCache_ is a different index built from `writeRegistry_` that can answer a different kind of query. I'll rename them for clarity.
- wc_: is unused and should be deleted.

File: torch/csrc/jit/ir/alias_analysis.cpp [hunk 12]
```diff
@@ -126,10 +126,78 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
...
+  auto& writeIndex = *writeIndex_;
+
+  // build the write index
+  for (const auto& write : writeRegistry_->writes_) {
+    auto it = elementMap_.find(write.second);
+    TORCH_INTERNAL_ASSERT(
+        it != elementMap_.end(), "Tried to write to value not in MemoryDAG");
+    const auto& writtenMemoryLocations =
+        memoryDAG_->getMemoryLocations(it->second);
+    writeIndex[write.first] |= writtenMemoryLocations;
```
[the remark this answers is not in the document]
Comment (suo) [reply torch/csrc/jit/ir/alias_analysis.cpp:174]: write index is a cache from node -> all memory locations it writes to. This for loop builds it up based on information we recorded during analysis. I'll rewrite the code for clarity

File: torch/csrc/jit/ir/alias_analysis.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/ir/alias_analysis.cpp:190]: If we analyzed `aten::fork` and `aten::wait` with `analyzeConservative` that would be correct, and would not require the above special casing leaking into memory dag

Comment (suo) [reply torch/csrc/jit/ir/alias_analysis.cpp:190]:
```
f : Future[None] = fork(a, b, c)  // mutates a
wait(f)
use(a)
```
it is illegal to reorder `wait` after `use`. `analyzeConservative` will only register a write to `f`, not `a`.

This can probably be improved. But I don't want to change behavior in this PR if possible.

File: torch/csrc/jit/ir/alias_analysis.cpp [hunk 13]
```diff
@@ -126,10 +126,78 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
...
+    void registerWrite(const Value* v, Node* n) {
+      writes_.emplace(n, v);
+    }
+    void registerWriteToAllContained(const Value* v, Node* n) {
+      containedWrites_.emplace(n, v);
+    }
+    void registerWriteToAllWildcards(Node* n) {
+      writesToAllWildcards_.insert(n);
+    }
+    std::unordered_map<Node*, const Value*> writes_;
```
[the remark this answers is not in the document]
Comment (suo) [reply torch/csrc/jit/ir/alias_analysis.cpp:141]: yes good catch, this is a bug. I'll fix. 

containedWrites are separate because it produces more restrictive results (e.g. appending to a list does not write to elements in the list, but it writes to list itself)

File: torch/csrc/jit/ir/alias_analysis.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/ir/alias_analysis.cpp:190]: `analyzeConservative` marks a write to all contained elements:
```
    auto elem = elementMap_.at(input);
    registerWrite(input, node);
    MemoryLocations mem_locations;
    memoryDAG_->collectAllContainedMemoryLocations(elem, mem_locations);
    for (unsigned loc : mem_locations) {
      auto contained_elem = memoryDAG_->fromIndex(loc);
      // we only register writes on memory locations
      if (contained_elem->pointsTo.empty()) {
        registerWrite(contained_elem, node);
      }
    }
```

File: torch/csrc/jit/ir/alias_analysis.cpp [hunk 14]
```diff
@@ -126,10 +126,81 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
 
 AliasDb::~AliasDb() = default;
 
+// Structure that keeps tracks of all writes at a high level. When alias
+// analysis is done, this will be used to construct a more efficient WriteIndex.
+struct AliasDb::WriteRegistry {
```
Comment (eellison) [inline torch/csrc/jit/ir/alias_analysis.cpp:131]: Maybe add comment somewhere that this exists before there is a value -> element mapping

File: torch/csrc/jit/ir/alias_analysis.cpp [hunk 15]
```diff
@@ -126,10 +126,81 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
...
+          it != elementMap_.end(), "Tried to write to value not in MemoryDAG");
+      const auto& writtenMemoryLocations =
+          memoryDAG_->getMemoryLocations(it->second);
+      writeIndex[node] |= writtenMemoryLocations;
+    }
+  }
+
+  for (const auto& write : writeRegistry_->containedWrites_) {
+    Node* node = write.first;
+    const std::vector<const Value*> writtenValues = write.second;
```
Comment (eellison) [inline torch/csrc/jit/ir/alias_analysis.cpp:184]: nit: probably want `const std::vector<const Value*>&` right ?

File: torch/csrc/jit/ir/alias_analysis.cpp [hunk 16]
```diff
@@ -126,10 +126,81 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
...
+      MemoryLocations writtenMemoryLocations;
+      memoryDAG_->collectAllContainedMemoryLocations(
+          elem, writtenMemoryLocations);
+      writeIndex[node] |= writtenMemoryLocations;
+    }
+  }
+
+  for (const auto& write : writeRegistry_->writesToAllWildcards_) {
+    for (const auto& pr : wildcardIndex_) {
+      writeIndex[write].set(pr.second->index);
```
Comment (eellison) [inline torch/csrc/jit/ir/alias_analysis.cpp:196]: are you sure this api call is right `writeIndex[write].set(pr.second->index)` ? Why are we using `.set()` here instead of `|=` as above ?

File: torch/csrc/jit/ir/alias_analysis.cpp [hunk 17]
```diff
@@ -126,10 +126,81 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
...
+    }
+  }
+
+  writeRegistry_ = nullptr; // to make future access an error
```
Comment (eellison) [inline torch/csrc/jit/ir/alias_analysis.cpp:200]: maybe add some more comments here

File: torch/csrc/jit/ir/alias_analysis.cpp [hunk 18]
```diff
@@ -779,17 +839,7 @@ void AliasDb::analyzeConservative(Node* node) {
...
-    MemoryLocations mem_locations;
-    memoryDAG_->collectAllContainedMemoryLocations(elem, mem_locations);
-    for (unsigned loc : mem_locations) {
-      auto contained_elem = memoryDAG_->fromIndex(loc);
-      // we only register writes on memory locations
-      if (contained_elem->pointsTo.empty()) {
-        registerWrite(contained_elem, node);
-      }
-    }
+    registerWrite(input, node, true);
```
Comment (eellison) [inline torch/csrc/jit/ir/alias_analysis.cpp:842]: nit can you add `/*writes_to_contained*/`  to the `true` arg.

File: torch/csrc/jit/passes/utils/memory_dag.cpp [hunk 19]
```diff
@@ -93,71 +107,101 @@ bool MemoryDAG::mayContainAlias(
...
-  return indexToElementMap_.back().get();
+Element* MemoryDAGBuilder::makeFreshValue(const Value* v) {
+  return makeFreshValueImpl(v, indexToElementMap_);
 }
 
-const MemoryLocations& Element::getMemoryLocations() const {
-  if (!cachedMemoryLocations_.empty()) {
-    return cachedMemoryLocations_;
+const MemoryLocations& MemoryDAG::getMemoryLocations(const Element* e) const {
+  if (e->cachedMemoryLocations_) {
```
Comment (eellison) [inline torch/csrc/jit/passes/utils/memory_dag.cpp:128]: nit: can you add comment to the caching here: when you should expect it to be set

File: torch/csrc/jit/passes/utils/memory_dag.cpp [hunk 20]
```diff
@@ -93,71 +107,101 @@ bool MemoryDAG::mayContainAlias(
...
-  ska::flat_hash_set<int> seen;
-  queue.push(this->index);
-  while (!queue.empty()) {
-    const auto index = queue.front();
-    queue.pop();
-    seen.insert(index);
-    auto el = dag.fromIndex(index);
-    if (el->pointsTo.empty()) {
-      res.set(index);
+  if (e->pointsTo.empty()) {
```
Comment (eellison) [inline torch/csrc/jit/passes/utils/memory_dag.cpp:133]: It's a little surprising that we're not using a `seen` set in our DFS implementation, it may not trigger now but that's an easy way for an exponential runtime. Could you add a `getMemoryLocationsImpl` with a seen set ?

**Review by eellison — APPROVED** [review]
This is a lot of complexity but given how much this is on the hot path i'm approving. Most of the complexity is because of wildcards and how they invalidate the memory dag state. We should investigate a context sensitive heap analysis so that wildcards are no longer special cased within the dag. 

Could you please address my comment in `getMemoryLocations` before landing ?

File: torch/csrc/jit/passes/utils/memory_dag.cpp (hunk 20 above)
Comment (suo) [reply torch/csrc/jit/passes/utils/memory_dag.cpp:133]: Not sure what a seen set would do? If we have traversed `e` before, its `cachedMemoryLocations_` will be populated and thus we will return immediately from the traversal.

File: torch/csrc/jit/ir/alias_analysis.cpp (hunk 16 above)
Comment (suo) [reply torch/csrc/jit/ir/alias_analysis.cpp:196]: |= is for adding many memorylocations. set() is for adding one.

File: torch/csrc/jit/ir/alias_analysis.cpp (hunk 15 above)
Comment (suo) [reply torch/csrc/jit/ir/alias_analysis.cpp:184]: good catch!

File: torch/csrc/jit/passes/utils/memory_dag.cpp (hunk 20 above)
Comment (eellison) [reply torch/csrc/jit/passes/utils/memory_dag.cpp:133]: yea nevermind, cachedMemoryLocations_ makes sure we don't redo work.
